app/src/main/java/com/example/hackernews/view/dialog/LoginDialog.kt

```diff
@@ -1,80 +1,66 @@
+class LoginDialog(private val activity: Activity) :
+    AppCompatDialogFragment() {
 
     private lateinit var binding: LoginLayoutBinding
+    private val loginViewModel: LoginViewModel by viewModels {
+        LoginViewModelFactory(requireContext())
+    }
 
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
         val builder = AlertDialog.Builder(activity)
         binding = LoginLayoutBinding.inflate(layoutInflater)
         initViews(builder)
-        binding.tvUsernameRequiredMessage.visibility = View.GONE
-        binding.tvPasswordRequiredMessage.visibility = View.GONE
-        builder.setPositiveButton("Register") { dialog, which ->
-            val username = Helper.trimEditText(binding.editUsername)
-            val password = Helper.trimEditText(binding.editPassword)
+        builder.setPositiveButton("Register") { _, _ ->
+            val username = binding.editUsername.text.toString()
+            val password = binding.editPassword.text.toString()
             registerUser(username, password)
         }
-        builder.setNegativeButton("Login") { dialog, which ->
-            val username = Helper.trimEditText(binding.editUsername)
-            val password = Helper.trimEditText(binding.editPassword)
-            val loggedUser = loginUser(username, password)
-            if(loggedUser != null) {
-                loggedUser.observe(this) {
-                    listener.onLoggedIn(username)
-                }
-            }
-            else {
-                listener.onLoggedFailed()
-                Log.e("DIALOG FAILED", "$loggedUser")
-            }
+        builder.setNegativeButton("Login") { _, _ ->
+            Toast.makeText(context, "Clicked on login", Toast.LENGTH_LONG).show()
         }
         return builder.create()
     }
 
-    private fun loginUser(username: String, password: String): LiveData<User?>? {
-        userViewModel.logUser(username, password)
-        return userViewModel.user
-    }
-
     private fun registerUser(username: String, password: String) {
-        val user: User = User(username, password, UUID.randomUUID().toString(), 0)
-        userViewModel.insert(user)
-        Toast.makeText(context, "Successfully registered $username", Toast.LENGTH_LONG).show()
+        val user = User(username, password, UUID.randomUUID().toString(), 1)
+        val executors = Executors.newSingleThreadExecutor()
+        executors.execute {
+            loginViewModel.registerUser(user, listener, requireContext())
+        }
     }
 
-    private fun validateForm(): Boolean {
-        if (binding.editUsername.text.toString().isEmpty()) {
-            binding.tvUsernameRequiredMessage.visibility = View.VISIBLE
-            return false
-        } else if (binding.editPassword.text.toString().isEmpty()) {
-            binding.tvPasswordRequiredMessage.visibility = View.VISIBLE
-            return false
+    private val listener: (Result) -> Unit = { result ->
```

Comment:
Use callback and do everything in ViewModel, only expose isLoggedIn observable property from ViewModel.

ViewModel code
```kotlin
val isRegistrationSuccess = MutableLiveData(false)

fun register(username, password) {
   // check errors
   // add user to database
   // update livedata
}

```